engine/crates/engine-v2/schema/src/input_value/raw.rs

```diff
@@ -0,0 +1,522 @@
+use crate::{
+    EnumValueId, IdRange, InputValueDefinitionId, RawInputKeyValueId, RawInputObjectFieldValueId, RawInputValueId,
+    SchemaInputValueId,
+};
+
+/// Holds input values for the Schema and for an Operation during execution.
+/// It's generic over 'Str', the string representation:
+/// - StringId for the Schema (default input values and directive arguments)
+/// - Box<str> for the Operation
+///
+/// This allow us to share the code for:
+/// - Display: used to print default values in introspection and add input values in subgraph query strings.
+/// - Serialize: used to serialize variables/arguments.
+/// - Deserializer: used by resolvers to deserialize arguments into a specific struct.
+///
+/// Data is stored in flat arrays for hopefully faster serialization and as as bonus smaller InputValue
+/// footprint on x64 architectures: u64 + usize (padding) rather than 3 * usize (Box<[]> + padding).
```

Comment:
The last bit of this docstring seems like it either needs corrected or clarified?  I can't see how 3 Vecs would only have a single u64 of footprint?  I'm also not clear on what the alternative that has `3 * usize` is?  

Also why a whole usize of padding?  That seems sus to me.

Reply:
Well interestingly Rust does some optimization when only a single `Vec` is present in the possibilities:

```rust
enum First {
    A(u64),
    B(Vec<u64>),
}

enum Second {
    A(u64),
    B(Vec<u64>),
    D(Vec<u128>)
}

struct IdRange<T> {
    start: T,
    end: T
}

enum Third {
    A(u64),
    B(IdRange<u32>),
    C(IdRange<u32>),
}

fn main() {
    println!("First size: {} align: {}", std::mem::size_of::<First>(), std::mem::align_of::<First>());
    println!("Second size: {} align: {}", std::mem::size_of::<Second>(), std::mem::align_of::<Second>());
    println!("Third size: {} align: {}", std::mem::size_of::<Third>(), std::mem::align_of::<Third>());
}
```
which outputs:
```text
First size: 24 align: 8
Second size: 32 align: 8
Third size: 16 align: 8
```

> I can't see how 3 Vecs would only have a single u64 of footprint?

With `IdRange` we're using two `u32` for start and end. We just need the enum variant tag + alignment padding. I may have extrapolated it to be `usize` though to be fair. I expect it to be a padding of `u32` for wasm32, but might be wrong.

Without `IdRange` we would have a least `Box<[T]>`  (`usize` for pointer + `usize` for len) + alignment padding. And obviously with the downside of having many small arrays rather than a single one